### PR TITLE
feat(catalog/glue): add support for glue catalog namespace operations

### PR DESCRIPTION
**Pull request by natusioe**
Continuation from [the previous PR](https://github.com/apache/iceberg-go/pull/169) to support Glue catalog. Add support for more namespace and table operations. In the next PR, remaining operations will be added to complete Glue catalog for tables (e.g. create table).

I mostly followed the same semantics as [the pyiceberg Glue catalog](https://github.com/apache/iceberg-python/blob/main/pyiceberg/catalog/glue.py) for the implementation. This means that Glue database fields (description and location URI) are stored as props. Error handling are also done in a similar way.

### Testing

- Unit tests are added. I am not sure if the project has AWS Glue set up (or some sort of realistic mock) for integration testing but I did not see it run as part of the CI actions.

Updated `cmd/iceberg/main.go` and manually tested using my personal AWS account:

```
> go run ./cmd/iceberg create --catalog glue --description "my new catalog" --location-uri "s3://test-namespace/" namespace apache-iceberg-test-00006

> go run ./cmd/iceberg describe --catalog glue namespace apache-iceberg-test-00006
┌──────────────────────────────────────┐
| Key           | Value                |
| ------------------------------------ |
| database_type | ICEBERG              |
| Description   | my new catalog       |
| Location      | s3://test-namespace/ |
└──────────────────────────────────────┘

> go run ./cmd/iceberg list --catalog glue
┌───────────────────────────┐
| IDs                       |
| ------------------------- |
| apache-iceberg-test-00001 |
| apache-iceberg-test-00002 |
| apache-iceberg-test-00003 |
| apache-iceberg-test-00004 |
| apache-iceberg-test-00005 |
| apache-iceberg-test-00006 |
└───────────────────────────┘

> go run ./cmd/iceberg properties --catalog glue get namespace apache-iceberg-test-00006 Description
my new catalog

> go run ./cmd/iceberg properties --catalog glue set namespace apache-iceberg-test-00006 Description "updated description"
updated Description on apache-iceberg-test-00006

> go run ./cmd/iceberg properties --catalog glue get namespace apache-iceberg-test-00006 Description
updated description

> go run ./cmd/iceberg describe --catalog glue namespace apache-iceberg-test-00006
┌──────────────────────────────────────┐
| Key           | Value                |
| ------------------------------------ |
| Description   | updated description  |
| Location      | s3://test-namespace/ |
| database_type | ICEBERG              |
└──────────────────────────────────────┘
```

### CODE REVIEW

File: catalog/glue.go [hunk 1]
```diff
@@ -30,22 +30,40 @@ import (
...
+	glueTypeIceberg      = "ICEBERG"
+	databaseTypePropsKey = "database_type"
+	tableTypePropsKey    = "table_type"
+	descriptionPropsKey  = "Description"
+
+	// Database location.
+	locationPropsKey = "Location"
+
+	// Table metadata location pointer.
+	metadataLocationPropsKey = "metadata_location"
```
Comment (zeroshade) [inline catalog/glue.go:45]: is there documentation that these came from? Can we link to the source documentation that these came from?

**Review by zeroshade — APPROVED** [review]
Generally LGTM just one question

File: catalog/glue.go (hunk 1 above)
Comment (natusioe) [reply catalog/glue.go:45]: I could not find the official spec but followed the same convention used by the PyIceberg [here](https://github.com/apache/iceberg-python/blob/main/pyiceberg/catalog/__init__.py#L88). Let me know if this works.

File: catalog/glue.go [hunk 2]
```diff
@@ -122,33 +149,93 @@ func (c *GlueCatalog) LoadTable(ctx context.Context, identifier table.Identifier
...
+		return err
+	}
+
+	// Check if the table exists and is an Iceberg table.
+	_, err = c.getTable(ctx, database, tableName)
+	if err != nil {
+		return err
+	}
+
+	params := &glue.DeleteTableInput{
```
Comment (nastra) [inline catalog/glue.go:181]: does this not need a glue catalog id?

Comment (nastra) [reply catalog/glue.go:181]: a glue catalog id is used in pyiceberg and in the java implementation, so I would have assumed that we're using it for the go client too

File: catalog/glue.go [hunk 3]
```diff
@@ -122,33 +149,93 @@ func (c *GlueCatalog) LoadTable(ctx context.Context, identifier table.Identifier
...
+	fromGlueTable, err := c.getTable(ctx, fromDatabase, fromTable)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch the table %s.%s: %w", fromDatabase, fromTable, err)
+	}
 
-func (c *GlueCatalog) UpdateNamespaceProperties(ctx context.Context, namespace table.Identifier,
-	removals []string, updates iceberg.Properties) (PropertiesUpdateSummary, error) {
-	return PropertiesUpdateSummary{}, fmt.Errorf("%w: [Glue Catalog] update namespace properties", iceberg.ErrNotImplemented)
+	// Create the new table.
+	_, err = c.glueSvc.CreateTable(ctx, &glue.CreateTableInput{
```
Comment (nastra) [inline catalog/glue.go:217]: I believe this is also missing the glue catalog id?

File: catalog/glue.go [hunk 4]
```diff
@@ -180,8 +267,155 @@ func (c *GlueCatalog) ListNamespaces(ctx context.Context, parent table.Identifie
...
+		databaseParameters[locationPropsKey] = locationURI
+	}
+
+	databaseInput := &types.DatabaseInput{
+		Name:       aws.String(database),
+		Parameters: databaseParameters,
+	}
+
+	params := &glue.CreateDatabaseInput{DatabaseInput: databaseInput}
+	_, err = c.glueSvc.CreateDatabase(ctx, params)
```
Comment (nastra) [inline catalog/glue.go:288]: this is also missing a glue catalog id

File: catalog/glue.go (hunk 1 above)
Comment (nastra) [reply catalog/glue.go:45]: I think you also need to set `previous_metadata_location` when committing: See https://github.com/apache/iceberg-python/blob/main/pyiceberg/catalog/glue.py#L142-L145

Comment (natusioe) [reply catalog/glue.go:45]: Added the documentation to the pyiceberg.

@nastra I will add `previous_metadata_location` in the next PR with the remaining ops (e.g. register_table) as it is used in them.

File: catalog/glue.go (hunk 2 above)
Comment (natusioe) [reply catalog/glue.go:181]: Added. Followed the Java pattern where we add catalog id in every request input. Event listener based approach from Python Iceberg might be ugly in Golang.

File: catalog/glue.go (hunk 3 above)
Comment (natusioe) [reply catalog/glue.go:217]: Yes, good catch, added.

File: catalog/glue.go (hunk 4 above)
Comment (natusioe) [reply catalog/glue.go:288]: Added.

File: catalog/glue.go [hunk 5]
```diff
@@ -122,38 +165,104 @@ func (c *GlueCatalog) LoadTable(ctx context.Context, identifier table.Identifier
...
 
-func (c *GlueCatalog) CreateNamespace(ctx context.Context, namespace table.Identifier, props iceberg.Properties) error {
-	return fmt.Errorf("%w: [Glue Catalog] create namespace", iceberg.ErrNotImplemented)
-}
+	toDatabase, toTable, err := identifierToGlueTable(to)
+	if err != nil {
+		return nil, err
+	}
 
-func (c *GlueCatalog) DropNamespace(ctx context.Context, namespace table.Identifier) error {
```
Comment (nastra) [inline catalog/glue.go:141]: moving methods around makes this quite difficult to review. I'd suggest to just add table-related methods after the namespace-related methods

Comment (natusioe) [reply catalog/glue.go:141]: Updated to same order now.

**Review by nastra — APPROVED** [review]

**Pull request merged**